File: MiniDOMWebCtrls/DMD/Forms/Charts/Renderers/VerticalBarOverlappedChartRenderer.cs

namespace minidom.Forms
{


    /// <summary>
    /// Renderer di un grafico a barre verticali
    /// </summary>
    /// <remarks></remarks>
    public class VerticalBarOverlappedChartRenderer : ChartRenderer
    {

        // Protected barWidth As Single
        protected float barDistance;
        // Private m_Subformat As BarChartFormats
        protected float m_BarWidth;

        public VerticalBarOverlappedChartRenderer(CChart chart) : base(chart)
        {
            // Me.barWidth = 10
            barDistance = 2f;
        }

        public override string Description
        {
            get
            {
                return "Vertical bars";
            }
        }

        public float BarWidth
        {
            get
            {
                return m_BarWidth;
            }
        }

        protected override void DrawSerie(CChartSerie serie)
        {
            float x;
            System.Drawing.Color color;
            System.Drawing.Brush brush;
            x = 0f;
            color = serie.GetRenderColor();
            brush = new System.Drawing.SolidBrush(color);
            foreach (CChartValue value in serie.Values)
            {
                FillRectangle(brush, x, 0f, BarWidth, (float)value.Value);
                DrawRectangle(System.Drawing.Pens.Black, x, 0f, BarWidth, (float)value.Value);
                x += BarWidth + barDistance;
            }

            brush.Dispose();
        }

        protected override void CalculateParameters()
        {
            bool init = false;
            int maxNumXSerie = 0;
            foreach (CChartSerie serie in Chart.Series)
            {
                var ms = serie.GetMeasures();
                if (init == false)
                {
                    minY = ms.MinY;
                    maxY = ms.MaxY;
                    init = true;
                }
                else
                {
                    if (ms.MinY < minY == true)
                        minY = ms.MinY;
                    if (ms.MaxY > maxY == true)
                        maxY = ms.MaxY;
                }

                maxNumXSerie = Maths.Max(serie.Values.Count, maxNumXSerie);
            }

            m_BarWidth = (float)(renderRect.Width / (double)(maxNumXSerie * Chart.Series.Count));
            minX = 0f;
            maxX = Chart.Labels.Count;
            foreach (CChartSerie serie in Chart.Series)
            {
                if (serie.Values.Count > maxX)
                    maxX = serie.Values.Count;
            }

            maxX = maxX * (BarWidth + barDistance);
            xScale = 1f;
            yScale = 1f;
            if (minY < maxY == true)
                yScale = (float)((renderRect.Height - 5) / (maxY - minY));
            // If (Me.minX < Me.maxX) Then Me.xScale = ((Me.renderRect.Width - 5) / (Me.maxX - Me.minX))
            xOrig = -minX;
            yOrig = (float)-minY;
        }

        public override ChartTypes Type
        {
            get
            {
                return ChartTypes.VerticalBarsOverlapped;
            }
        }
    }
}

File: MiniDOMWebCtrls/DMD/Forms/Charts/Renderers/VerticalBarOverlappedChartRenderer.vb
Imports minidom
Imports minidom.Sistema
Imports minidom.Forms


Namespace Forms
 

    ''' <summary>
    ''' Renderer di un grafico a barre verticali
    ''' </summary>
    ''' <remarks></remarks>
    Public Class VerticalBarOverlappedChartRenderer
        Inherits ChartRenderer

        'Protected barWidth As Single
        Protected barDistance As Single
        'Private m_Subformat As BarChartFormats
        Protected m_BarWidth As Single


        Public Sub New(ByVal chart As CChart)
            MyBase.New(chart)
            '   Me.barWidth = 10
            Me.barDistance = 2
        End Sub

        Public Overrides ReadOnly Property Description As String
            Get
                Return "Vertical bars"
            End Get
        End Property

        Public ReadOnly Property BarWidth As Single
            Get
                Return Me.m_BarWidth
            End Get
        End Property



        Protected Overrides Sub DrawSerie(ByVal serie As CChartSerie)
            Dim x As Single
            Dim color As System.Drawing.Color
            Dim brush As System.Drawing.Brush
            x = 0
            color = serie.GetRenderColor
            brush = New System.Drawing.SolidBrush(color)
            For Each value As CChartValue In serie.Values
                Me.FillRectangle(brush, x, 0, Me.BarWidth, value.Value)
                Me.DrawRectangle(Drawing.Pens.Black, x, 0, Me.BarWidth, value.Value)
                x += Me.BarWidth + Me.barDistance
            Next
            brush.Dispose()
        End Sub

        Protected Overrides Sub CalculateParameters()
            Dim init As Boolean = False
            Dim maxNumXSerie As Integer = 0

            For Each serie As CChartSerie In Me.Chart.Series
                Dim ms As ChartMeasure = serie.GetMeasures
                If (init = False) Then
                    Me.minY = ms.MinY
                    Me.maxY = ms.MaxY
                    init = True
                Else
                    If (ms.MinY < Me.minY) Then Me.minY = ms.MinY
                    If (ms.MaxY > Me.maxY) Then Me.maxY = ms.MaxY
                End If
                maxNumXSerie = Math.Max(serie.Values.Count, maxNumXSerie)
            Next

            Me.m_BarWidth = Me.renderRect.Width / (maxNumXSerie * Me.Chart.Series.Count)

            Me.minX = 0
            Me.maxX = Me.Chart.Labels.Count
            For Each serie As CChartSerie In Me.Chart.Series
                If serie.Values.Count > Me.maxX Then Me.maxX = serie.Values.Count
            Next
            Me.maxX = Me.maxX * (Me.barWidth + Me.barDistance)
            Me.xScale = 1 : Me.yScale = 1
            If (Me.minY < Me.maxY) Then Me.yScale = ((Me.renderRect.Height - 5) / (Me.maxY - Me.minY))
            'If (Me.minX < Me.maxX) Then Me.xScale = ((Me.renderRect.Width - 5) / (Me.maxX - Me.minX))
            Me.xOrig = -Me.minX
            Me.yOrig = -Me.minY
        End Sub

        Public Overrides ReadOnly Property Type As ChartTypes
            Get
                Return ChartTypes.VerticalBarsOverlapped
            End Get
        End Property



    End Class

End Namespace